File: src/main/java/com/zerozero/core/util/JwtUtil.java
package com.zerozero.core.util;

import com.zerozero.auth.exception.AuthenticationErrorCode;
import com.zerozero.core.domain.entity.User;
import com.zerozero.core.domain.vo.AccessToken;
import com.zerozero.core.domain.vo.RefreshToken;
import io.jsonwebtoken.Claims;
import io.jsonwebtoken.Jwts;
import io.jsonwebtoken.SignatureAlgorithm;
import io.jsonwebtoken.io.Decoders;
import io.jsonwebtoken.security.Keys;
import java.security.Key;
import java.util.Date;
import java.util.Map;
import java.util.UUID;
import java.util.function.Function;
import lombok.RequiredArgsConstructor;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.stereotype.Service;

@Service
@RequiredArgsConstructor
public class JwtUtil {

  @Value("${jwt.secretKey}")
  private String secretKey;

  @Value("${jwt.expiration}")
  private long jwtExpiration;

  @Value("${jwt.refresh-token.expiration}")
  private long refreshExpiration;

  public UUID extractUserId(String token) {
    return UUID.fromString(extractClaim(token, claims -> claims.get("userId", String.class)));
  }

  public <T> T extractClaim(String token, Function<Claims, T> claimsResolver) {
    final Claims claims = extractAllClaims(token);
    return claimsResolver.apply(claims);
  }

  public AccessToken generateAccessToken(User user) {
    String accessToken = Jwts.builder()
        .setClaims(Map.of(
            "userId", user.getId(),
            "email", user.getEmail()
        ))
        .setIssuedAt(new Date(System.currentTimeMillis()))
        .setExpiration(new Date(System.currentTimeMillis() + jwtExpiration))
        .signWith(getSignInKey(), SignatureAlgorithm.HS256)
        .compact();
    return AccessToken.of(accessToken);
  }

  public RefreshToken generateRefreshToken(User user) {
    String refreshToken = Jwts.builder()
        .setClaims(Map.of(
            "userId", user.getId(),
            "email", user.getEmail()
        ))
        .setIssuedAt(new Date(System.currentTimeMillis()))
        .setExpiration(new Date(System.currentTimeMillis() + refreshExpiration))
        .signWith(getSignInKey(), SignatureAlgorithm.HS256)
        .compact();
    return RefreshToken.of(refreshToken);
  }

  public void validateExpiration(String token) {
    if (isTokenExpired(token)) {
      throw AuthenticationErrorCode.EXPIRED_TOKEN.toException();
    }
  }

  public boolean isTokenValid(String token, User user) {
    final UUID userId = extractUserId(token);
    return (userId.equals(user.getId())) && !isTokenExpired(token);
  }

  public boolean isTokenExpired(String token) {
    return extractExpiration(token).before(new Date());
  }

  private Date extractExpiration(String token) {
    return extractClaim(token, Claims::getExpiration);
  }

  private Claims extractAllClaims(String token) {
    try {
      return Jwts
          .parserBuilder()
          .setSigningKey(getSignInKey())
          .build()
          .parseClaimsJws(token)
          .getBody();
    } catch (Exception e) {
      throw AuthenticationErrorCode.NOT_DEFINE_TOKEN.toException();
    }
  }

  private Key getSignInKey() {
    byte[] keyBytes = Decoders.BASE64.decode(secretKey);
    return Keys.hmacShaKeyFor(keyBytes);
  }
}

File: src/main/java/com/zerozero/core/util/JwtUtil.kt
package com.zerozero.core.util

import com.zerozero.auth.exception.AuthErrorType
import com.zerozero.auth.exception.AuthException
import com.zerozero.user.domain.model.User
import io.jsonwebtoken.Claims
import io.jsonwebtoken.Jwts
import io.jsonwebtoken.SignatureAlgorithm
import io.jsonwebtoken.io.Decoders
import io.jsonwebtoken.security.Keys
import org.springframework.beans.factory.annotation.Value
import org.springframework.stereotype.Service
import java.security.Key
import java.util.*
import java.util.function.Function

@Service
class JwtUtil(
    @Value("\${jwt.secretKey}")
    private val secretKey: String,

    @Value("\${jwt.expiration}")
    private val jwtExpiration: Long,

    @Value("\${jwt.refresh-token.expiration}")
    private val refreshExpiration: Long
) {

    fun extractUserId(token: String): UUID {
        return UUID.fromString(extractClaim(token) { claims -> claims.get("userId", String::class.java) })
    }

    fun <T> extractClaim(token: String, claimsResolver: Function<Claims, T>): T {
        val claims = extractAllClaims(token)
        return claimsResolver.apply(claims)
    }

    fun generateAccessToken(user: User): String {
        return Jwts.builder()
            .setClaims(
                mapOf(
                    "userId" to user.id,
                    "email" to user.email
                )
            )
            .setIssuedAt(Date(System.currentTimeMillis()))
            .setExpiration(Date(System.currentTimeMillis() + jwtExpiration))
            .signWith(signInKey(), SignatureAlgorithm.HS256)
            .compact()
    }

    fun generateRefreshToken(user: User): String {
        return Jwts.builder()
            .setClaims(
                mapOf(
                    "userId" to user.id,
                    "email" to user.email
                )
            )
            .setIssuedAt(Date(System.currentTimeMillis()))
            .setExpiration(Date(System.currentTimeMillis() + refreshExpiration))
            .signWith(signInKey(), SignatureAlgorithm.HS256)
            .compact()
    }

    fun validateExpiration(token: String) {
        if (isTokenExpired(token)) {
            throw AuthException(AuthErrorType.EXPIRED_TOKEN)
        }
    }

    fun isTokenValid(token: String, user: User): Boolean {
        val userId = extractUserId(token)
        return (userId == user.id) && !isTokenExpired(token)
    }

    fun isTokenExpired(token: String): Boolean {
        return extractExpiration(token).before(Date())
    }

    private fun extractExpiration(token: String): Date {
        return extractClaim(token) { obj -> obj.expiration }
    }

    private fun extractAllClaims(token: String): Claims {
        try {
            return Jwts
                .parserBuilder()
                .setSigningKey(signInKey())
                .build()
                .parseClaimsJws(token)
                .body
        } catch (e: Exception) {
            throw AuthException(AuthErrorType.NOT_DEFINE_TOKEN)
        }
    }

    private fun signInKey(): Key {
        val keyBytes = Decoders.BASE64.decode(secretKey)
        return Keys.hmacShaKeyFor(keyBytes)
    }
}
